[Non_Sanity] No identity resp with mme restart shows flakiness

### Your Environment

- **Version:** Local Development Setup
- **Affected Component:** Access Gateway
- **Affected Subcomponent:** Magma MME/SPGW service
- **Deployment Environment:** Vagrant (AGW)

### Setup Details
The S1APTester ([https://github.com/facebookexperimental/S1APTester](https://github.com/facebookexperimental/S1APTester)) was used along with Magma EPC in Vagrant based setup to reproduce the issue.

### Describe the Issue
The non-sanity test case s1aptests/test_no_identity_rsp_with_mme_restart.py is showing flaky behavior. Either we run this test case multiple times in loop or we place the test case after some test cases from top in the list of non sanity test cases, this test case fails many time with assertion failure. This issue is similar to the issue #9519 where MME restart procedure is taking extra time for some executions, causing the identification timer T3470 to expire before MME could restart. This forces MME to re-transmit identity request message, but the test case was expecting UE context release indication and hence leads to assertion failure. The relevant logs and pcaket captures are attached here:
[test_no_identity_rsp_with_mme_restart.py.zip](https://github.com/magma/magma/files/7419717/test_no_identity_rsp_with_mme_restart.py.zip)



Following is the terminal log for failure scenario:
```
. /home/vagrant/build/python/bin/activate
echo "Running test: s1aptests/test_no_identity_rsp_with_mme_restart.py"
Running test: s1aptests/test_no_identity_rsp_with_mme_restart.py
timeout --foreground -k 930s 900s sudo -E PATH=/home/vagrant/build/python/bin:/usr/lib/ccache:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/go/bin PYTHONPATH=:/home/vagrant/s1ap-tester/bin /home/vagrant/build/python/bin/nosetests --with-xunit --xunit-file=/var/tmp/test_results/test_no_identity_rsp_with_mme_restart.xml -x -s s1aptests/test_no_identity_rsp_with_mme_restart.py || exit 1
Start time 15:34:21
tag: AUTH_TYPE tagVal: MILENAGE
************************* Enb tester config

**********************TPT open server******** 0

 Server Socket FD =[13]
************* Testing *Time elapsed(in usec): for timer expiry               :9094 usec
AGW is stateless
APN Correction configured
Health service is disabled
Using subscriber IMSI IMSI001010000000001
Using IMEI 3805468432113171
************************* UE device config for ue_id  1
************************* Configuring IP block
************************* Waiting for IP changes to propagate
************************* S1 setup
************************* UE App config
************************* Running No Identity Response with MME   restart for UE id  1
Sending Attach Request for ue_id 1
Received Identity req ind  17
********************** Restarting MME service on gateway ***
Waiting for 0 seconds
Waiting for 1 seconds
Waiting for 2 seconds
Waiting for 3 seconds
Waiting for 4 seconds
Waiting for 5 seconds
Waiting for 6 seconds
Waiting for 7 seconds
Waiting for 8 seconds
Waiting for 9 seconds
Waiting for 10 seconds
Waiting for 11 seconds
Waiting for 12 seconds
Waiting for 13 seconds
Waiting for 14 seconds
Waiting for 15 seconds
Waiting for 16 seconds
Waiting for 17 seconds
Waiting for 18 seconds
Waiting for 19 seconds
Waiting for 20 seconds
Waiting for 21 seconds
Waiting for 22 seconds
Waiting for 23 seconds
Waiting for 24 seconds
Waiting for 25 seconds
Waiting for 26 seconds
Waiting for 27 seconds
Waiting for 28 seconds
Waiting for 29 seconds
Sending Attach Request for ue_id 1
************************* send SCTP SHUTDOWN
Keys left in Redis (list should be empty)[

]
Entries in s1ap_imsi_map (should be zero): 0
Entries left in hashtables (should be zero): 0
Entries in mme_ueip_imsi_map (should be zero): 0
F
======================================================================
FAIL: Step 1: UE performs GUTI based attach
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vagrant/magma/lte/gateway/python/integ_tests/s1aptests/test_no_identity_rsp_with_mme_restart.py", line 104, in test_no_identity_rsp_with_mme_restart
    response.msg_type, s1ap_types.tfwCmd.UE_CTX_REL_IND.value,
AssertionError: 17 != 52
-------------------- >> begin captured logging << --------------------
root: ERROR: Could not load magmad yml config for mconfig modules
urllib3.connectionpool: DEBUG: Starting new HTTP connection (1): 192.168.60.142:3333
urllib3.connectionpool: DEBUG: http://192.168.60.142:3333 "GET /stats/switches HTTP/1.1" 200 17
urllib3.connectionpool: DEBUG: Starting new HTTP connection (1): 192.168.60.142:3333
urllib3.connectionpool: DEBUG: http://192.168.60.142:3333 "GET /stats/switches HTTP/1.1" 200 17
root: INFO: Adding subscriber : IMSI001010000000001
root: INFO: Deleting subscriber : IMSI001010000000001
--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------
Ran 1 test in 42.057s

FAILED (failures=1)
```

This is an extension of the issue #5842 

**To Reproduce**
1) Run the test case in loop "make integ_test TESTS=s1aptests/test_no_identity_rsp_with_mme_restart.py"

--- a/lte/gateway/python/integ_tests/s1aptests/test_no_identity_rsp_with_mme_restart.py
+++ b/lte/gateway/python/integ_tests/s1aptests/test_no_identity_rsp_with_mme_restart.py
@@ -20,12 +20,16 @@ from s1ap_utils import MagmadUtil
 
 
 class TestNoIdentityRspWithMmeRestart(unittest.TestCase):
+    """Unittest: TestNoIdentityRspWithMmeRestart"""
+
     def setUp(self):
+        """Initialize before test case execution"""
         self._s1ap_wrapper = s1ap_wrapper.TestWrapper(
             stateless_mode=MagmadUtil.stateless_cmds.ENABLE,
         )
 
     def tearDown(self):
+        """Cleanup after test case execution"""
         self._s1ap_wrapper.cleanup()
 
     def test_no_identity_rsp_with_mme_restart(self):
@@ -56,12 +60,14 @@ class TestNoIdentityRspWithMmeRestart(unittest.TestCase):
         attach_req.useOldSecCtxt = sec_ctxt
         print("Sending Attach Request for ue_id", req.ue_id)
         self._s1ap_wrapper._s1_util.issue_cmd(
-            s1ap_types.tfwCmd.UE_ATTACH_REQUEST, attach_req,
+            s1ap_types.tfwCmd.UE_ATTACH_REQUEST,
+            attach_req,
         )
 
         response = self._s1ap_wrapper.s1_util.get_response()
         self.assertEqual(
-            response.msg_type, s1ap_types.tfwCmd.UE_IDENTITY_REQ_IND.value,
+            response.msg_type,
+            s1ap_types.tfwCmd.UE_IDENTITY_REQ_IND.value,
         )
         print(
             "Received Identity req ind ",
@@ -75,16 +81,14 @@ class TestNoIdentityRspWithMmeRestart(unittest.TestCase):
             print("Waiting for", j, "seconds")
             time.sleep(1)
 
-        """
-         Since UE has neither received Attach Reject nor Attach Accept,
-         assuming that both T3410 and T3411 timer expires at UE
-         and re-transmit the Attach Request with same IEs in UL nas message
-         with previous mme_ue_s1ap_id
-         But in response to second Attach Request, s1ap receives ue context
-         release command because after AGW recovery, s1ap stack will not be
-         aware of mme_ue_s1ap_id.So s1ap stack should clear the s1ap signaling
-         contexts at eNB that were created before mme restarts
-        """
+        # Since UE has neither received Attach Reject nor Attach Accept,
+        # assuming that both T3410 and T3411 timer expires at UE
+        # and re-transmit the Attach Request with same IEs in UL nas message
+        # with previous mme_ue_s1ap_id
+        # But in response to second Attach Request, s1ap receives ue context
+        # release command because after AGW recovery, s1ap stack will not be
+        # aware of mme_ue_s1ap_id.So s1ap stack should clear the s1ap signaling
+        # contexts at eNB that were created before mme restarts
         attach_req = s1ap_types.ueAttachRequest_t()
         attach_req.ue_Id = req.ue_id
         sec_ctxt = s1ap_types.TFW_CREATE_NEW_SECURITY_CONTEXT
@@ -95,13 +99,36 @@ class TestNoIdentityRspWithMmeRestart(unittest.TestCase):
         attach_req.useOldSecCtxt = sec_ctxt
         print("Sending Attach Request for ue_id", req.ue_id)
         self._s1ap_wrapper._s1_util.issue_cmd(
-            s1ap_types.tfwCmd.UE_ATTACH_REQUEST, attach_req,
+            s1ap_types.tfwCmd.UE_ATTACH_REQUEST,
+            attach_req,
         )
 
+        # It has been observed that despite getting the restart command on
+        # time, MME sometimes restarts after a delay of around 6 seconds.
+        # Currently identity response timer T3470 is set to 6 seconds as
+        # part of configuraton file mme.conf.template. If MME restarts after
+        # expiry of identity response timer, it will send the re-transmitted
+        # identity request message
+        resp_count = 0
+        while True:
+            response = self._s1ap_wrapper.s1_util.get_response()
+            if (
+                response.msg_type
+                == s1ap_types.tfwCmd.UE_IDENTITY_REQ_IND.value
+            ):
+                resp_count += 1
+                print(
+                    "******************** Ignoring re-transmitted (",
+                    resp_count,
+                    ") Identity request indication",
+                )
+            else:
+                break
+
         # Context release
-        response = self._s1ap_wrapper.s1_util.get_response()
         self.assertEqual(
-            response.msg_type, s1ap_types.tfwCmd.UE_CTX_REL_IND.value,
+            response.msg_type,
+            s1ap_types.tfwCmd.UE_CTX_REL_IND.value,
         )
         print("********** UE Context released **********")
 
@@ -126,7 +153,6 @@ class TestNoIdentityRspWithMmeRestart(unittest.TestCase):
         self._s1ap_wrapper.s1_util.detach(
             req.ue_id,
             s1ap_types.ueDetachType_t.UE_SWITCHOFF_DETACH.value,
-            False,
         )
 
 

